BUG: Ensure df.itertuples() uses plain tuples correctly

simongibbons opened the pull request:
Currently DataFrame.itertuples() has an off by one error
when it inspects whether or not it should return namedtuples
or plain tuples in it's response.

This PR addresses that bug by correcting the condition
that is used when making the check.

Closes: #28282

- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

simongibbons committed:
BUG: Ensure df.itertuples() uses plain tuples correctly

Currently DataFrame.itertuples() has an off by one error
when it inspects whether or not it should return namedtuples
or plain tuples in it's response.

This PR addresses that bug by correcting the condition
that is used when making the check.

Closes: #28282

simongibbons commented on pandas/tests/frame/test_api.py:
```diff
@@ -288,6 +288,22 @@ def test_sequence_like_with_categorical(self):
+
+        # Dataframes with >=255 columns will fallback to regular tuples
+        with pytest.raises(AttributeError):
+            result_255_columns.foo_1
```
Because `namedtuple` generates a new class when it is used, the only way of determining if something is a namedtuple or a plain one is to try and use it as such.

Hopefully you think this is sufficiently clear with the comment.

jreback commented on pandas/core/frame.py:
```diff
@@ -1018,8 +1018,8 @@ def itertuples(self, index=True, name="Pandas"):
 
-        # Python 3 supports at most 255 arguments to constructor
-        if name is not None and len(self.columns) + index < 256:
+        # Python versions before 3.7 support at most 255 arguments to constructor
```
is it possible to skip the arg len check for >= 3.7? does it work?

simongibbons replied:
yes, this was simple to do.

jreback commented on pandas/tests/frame/test_api.py:
```diff
@@ -288,6 +288,22 @@ def test_sequence_like_with_categorical(self):
         for c, col in df.items():
             str(s)
 
+    def test_itertuples_fallback_to_regular_tuples(self):
```
can you move this next to the other itertuples tests

simongibbons replied:
Done

simongibbons committed:
Address comments.

1. Ensure we return named tuples in more cases (when using python >=
   3.7)
2. Move test around to be with the itertuples test
3. Update docstring with the new behaviour.

simongibbons commented on pandas/tests/frame/test_api.py:
```diff
@@ -261,8 +263,27 @@ def test_itertuples(self, float_frame):
-        assert not hasattr(tup3, "_fields")
         assert isinstance(tup3, tuple)
+        if PY37:
+            assert hasattr(tup3, "_fields")
```
This test has changed as we will return named tuples always for python >= 3.7 now.

WillAyd reviewed (approved):
lgtm

jreback commented:
thanks @simongibbons very nice!
